ams/server/src/main/java/com/netease/arctic/server/optimizing/OptimizingQueue.java

```diff
@@ -330,75 +216,105 @@ public void updateOptimizerGroup(ResourceGroup optimizerGroup) {
-        AmoroTable<?> table = tableManager.loadTable(tableRuntime.getTableIdentifier());
-        OptimizingPlanner planner =
-            new OptimizingPlanner(
-                tableRuntime.refresh(table),
-                (ArcticTable) table.originalTable(),
-                getAvailableCore());
-        if (tableRuntime.isBlocked(BlockableOperation.OPTIMIZE)) {
-          LOG.info("{} optimize is blocked, continue", tableRuntime.getTableIdentifier());
-          continue;
-        }
-        plannedTables.add(table.id());
-        if (planner.isNecessary()) {
-          TableOptimizingProcess optimizingProcess = new TableOptimizingProcess(planner);
-          LOG.info(
-              "{} after plan get {} tasks",
-              tableRuntime.getTableIdentifier(),
-              optimizingProcess.getTaskMap().size());
-          optimizingProcess.taskMap.values().forEach(taskQueue::offer);
-          break;
-        } else {
-          tableRuntime.cleanPendingInput();
-        }
-      } catch (Throwable e) {
-        LOG.error(tableRuntime.getTableIdentifier() + " plan failed, continue", e);
-      }
+  }
+
+  private void triggerAsyncPlanning(
+      TableRuntime tableRuntime, Set<TableRuntime> skipTables, long startTime) {
+    LOG.info("Trigger planning table {}", tableRuntime.getTableIdentifier());
+    planningTables.add(tableRuntime);
+    doPlanning(tableRuntime)
+        .whenComplete(
+            (process, throwable) -> {
+              long currentTime = System.currentTimeMillis();
+              scheduleLock.lock();
+              try {
+                plannedKeepingTables.put(tableRuntime, System.currentTimeMillis());
+                planningTables.remove(tableRuntime);
+                if (process != null) {
+                  tableQueue.offer(process);
+                  LOG.info(
+                      "{} completes planning {} tasks with a total cost of {} ms, skipping tables, {}",
+                      optimizerGroup.getName(),
+                      process.getTaskMap().size(),
+                      currentTime - startTime,
+                      skipTables);
+                } else {
+                  if (throwable != null) {
+                    LOG.error(
+                        "Planning table {} failed", tableRuntime.getTableIdentifier(), throwable);
+                  }
+                  LOG.info(
+                      "{} skip planning table {} with a total cost of {} ms.",
+                      optimizerGroup.getName(),
+                      tableRuntime.getTableIdentifier(),
+                      currentTime - startTime);
+                }
+                planningCompleted.signalAll();
+              } finally {
+                scheduleLock.unlock();
+              }
+            });
+  }
+
+  private CompletableFuture<TableOptimizingProcess> doPlanning(TableRuntime tableRuntime) {
+    CompletableFuture<TableOptimizingProcess> future = new CompletableFuture<>();
+    planExecutor.execute(
```

Comment:
The Future did not perceive the exceptional situation. You can using 
```
 future = CompletableFuture.supplyAsync(() -> xxx, planExecutor);
```
or Add
```
future.completeExceptionally(xxx)
```
in Catch statement

Reply:
good idea